### DISCUSSION TITLE
Auto generating selector hooks is not linted by `eslint-plugin-react-hooks`

### DISCUSSION THREAD

**Discussion by lenwhite** [pmndrs/zustand#1243] [General]
Referring to the documentation recommendation on [Auto Generating Selectors](https://docs.pmnd.rs/zustand/guides/auto-generating-selectors): while the pattern is clean, it also unfortunately disregards linting for rules of hooks via `eslint-plugin-react-hooks`.

The example:

```typescript
// get the property
const bears = useBearStore.use.bears()
```

would not be linted as a hook normally would be (i.e. no warning if its accidentally used after an early return, conditionally, outside a component, etc.)

I'm not sure if there's any elegant solution to this. 

--- 

Given that the linting rule does check for hooks if they're namespaced under a PascalCased object, something like this could work: 

```TypeScript
export type Selectors<S> = S extends { getState: () => infer T }
  ? { [K in keyof T & string as `use${Capitalize<K>}`]: () => T[K] }
  : never;

export const createSelectors = <T, S extends UseBoundStore<StoreApi<T>>>(store: S) => {
  return Object.keys(store.getState()).reduce((acc, k) => {
    return {
      ...acc,
      [`use${k[0].toUpperCase()}${k.slice(1)}`]: store((s) => s[k as keyof typeof s]),
    };
  }, {} as Selectors<S>);
};
```

Which can be used as: 

```TypeScript
SelectBear = createSelectors(useBearStore);
...
const bears = SelectBear.useBears();
```

But subjective as aesthetics may be, this seems much worse compared to the normal form of `useBearStore(state => state.bears)`

Marginally better:

```TypeScript
const createSlices = <T, S extends UseBoundStore<StoreApi<T>>>(store: S) => {
  return Object.keys(store.getState()).reduce((acc, k) => {
    return {
      ...acc,
      [k]: () => store((s) => s[k as keyof typeof s]),
    };
  }, {});
};
...
const slices = createSlices(useBearStore);
const useBearSlice = () => slices;
...
const bears = useBearSlice().bear();
```

Adding a proxy:

```TypeScript

export const createSlices = <ST, S extends UseBoundStore<StoreApi<ST>>>(store: S) => {
  const slices = Object.keys(store.getState()).reduce((acc, k) => {
    return {
      ...acc,
      [k]: () => store((s) => s[k as keyof typeof s]),
    };
  }, {} as S extends { getState: () => infer T } ? { [K in keyof T]: () => T[K] } : never);
  return new Proxy(slices, {
    get: (target, prop) => {
      if (Reflect.has(target, prop)) {
        return target[prop as keyof typeof target]();
      }
      return undefined;
    },
  }) as unknown as S extends { getState: () => infer T } ? { [K in keyof T]: T[K] } : never;
};
...
const bears = useBearSlice().bear;
```

**Comment by dai-shi** [comment]
For TS friendliness, I would personally prefer something like: `useBearStore.useSlice('bears')` (which translates to `useBearStore(state => state.bears)`)

**Reply by lenwhite** [reply]
An attribute of `useBearStore` isn't linted as a hook, even if it were in the format `useSlice`. To be linted as a hook per the [linting rules](https://github.com/facebook/react/blob/main/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js), it'll have to be in the format `useFoo()` or `PascalCase.useFoo()`.

`useBearSlice('bears')` would work, though.

**Reply by dai-shi** [reply]
> PascalCase.useFoo()

Didn't know that...

**Reply by lenwhite** [reply]
Is there a reason why we can't overload `useBearStore` and allow path selectors like `useBearStore('bears')`?

**Reply by dai-shi** [reply]
You can't overload a function in JS, but you can redefine. So, yeah, probably like this?
```js
const useBearStore = createCustomHook(create(...))
```

**Reply by lenwhite** [reply]
Sounds like it'll work, but I'm not good enough at TypeScript to properly type that along with the accompanying attributes on the store.
